### ISSUE TITLE
issue with small and wide object detection 

### ISSUE THREAD

**Issue by Mann1904** [ultralytics/yolov5#8391]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

![train_batch0 (copy)](https://user-images.githubusercontent.com/58176237/176406080-ce042d40-4fd9-4db0-aa2f-ba5f3a6ea385.jpg)

![train_batch0 (copy)(1)](https://user-images.githubusercontent.com/58176237/176410236-2115edcf-940a-45b5-b5b8-ec3393e94990.jpg)

This is a ground truth augmented image that was saved during training. 

Why was the bounding box of 3 rows (marked by a yellow arrow) missing in this image?

**Comment by glenn-jocher** [comment]
@Mann1904 high aspect ratio labels or very narrow labels may be rejected by candidate function here:
https://github.com/ultralytics/yolov5/blob/6935a54e603d634f6b0a9026604dc5875d1ca990/utils/augmentations.py#L279-L284

**Comment by Mann1904** [comment]
Why does the candidate function reject high aspect ratio labels or very narrow labels?

**Comment by glenn-jocher** [comment]
@Mann1904 it's tuned for the most common types of datasets like COCO. You're free to modify as you see fit for your custom data.

**Comment by Mann1904** [comment]
Thank you! @glenn-jocher

**Comment by mbenami** [comment]
Hi @Mann1904 
I got a similar issue with my dataset 
did you solve yours? 
can you share how?
I tried to use the autoanchor but didn't got good results 
Thanks

**Comment by glenn-jocher** [comment]
@mbenami you can always manually supply anchors in a model yaml and then point to it with the train.py --cfg argument, or to simply use the default anchors and disable autoanchor you can use python train.py `--noautoanchor`
https://github.com/ultralytics/yolov5/blob/b32f67f6beb4a921c98301fe7724003e23103728/train.py#L446

**Comment by mbenami** [comment]
@glenn-jocher 
1. I thought about setting the anchor manually but then I will lose the transfer learning from the pre-train ckpt no?
2. for changing manually you mean here? 
https://github.com/ultralytics/yolov5/blob/7398d2d77cbac9f66259926d49c26bfa3c257a9b/models/yolov5n.yaml#L7
if so there is some rule/ tutorial/ explanation on how to do that and which value should I choose 
3. I run the train without `--noautoanchor` which run the GE algo to find better anchors but still didn't get good results 
4. can the anchor be learned with hyperparameter tune?

Thanks for the help

**Comment by glenn-jocher** [comment]
@mbenami 👋 Hello! Thanks for asking about model anchors. YOLOv5 🚀 uses a new [Ultralytics](https://ultralytics.com/) algorithm called **AutoAnchor** for anchor verification and generation before training starts.

Autoanchor will analyse your anchors against your dataset and training settings (like `--img-size`), and will adjust your anchors as necessary if it determines the original anchors are a poor fit, or if an **anchor count** was specified in your model.yaml rather than **anchor values**, i.e. 
```yaml
# Specify anchor count (per layer)
anchors: 3

# --OR-- Specify anchor values manually
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
```

When generating new anchors, autoanchor first applies a kmeans function against your dataset labels (scaled to your training `--img-size`), and uses kmeans centroids as initial conditions for a [Genetic Evolution](https://en.wikipedia.org/wiki/Genetic_algorithm) (GE) algorithm. The GE algorithm will evolve all anchors for 1000 generations under default settings, using CIoU loss (same regression loss used during training) combined with [Best Possible Recall](https://en.wikipedia.org/wiki/Precision_and_recall) (BPR) as its fitness function.

Notebook example: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img width="1200" alt="Screenshot 2022-03-24 at 10 58 28" src="https://user-images.githubusercontent.com/26833433/159891298-8b15f80b-f5b0-4fa9-8824-7380bae32a76.png">

**No action is required on your part** to use autoanchor. If you would like to force manual anchors for any reason, you can skip autoanchor with the `--noautoanchor` flag:
```shell
python train.py --noautoanchor
```

For more details on AutoAnchor see:
https://github.com/ultralytics/yolov5/blob/master/utils/autoanchor.py

Good luck 🍀 and let us know if you have any other questions!
